File: src/common/py/api/metadata/metadata_commands.py
import dataclasses

from ...core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ...core.messaging.composers import (
    CommandComposer,
    CommandReplyComposer,
    MessageBuilder,
)
from ...data.entities.metadata import MetadataProfileContainerList


@Message.define("command/metadata/profiles")
class GetMetadataProfilesCommand(Command):
    """
    Command to fetch all global metadata profiles.

    Notes:
        Requires a ``GetMetadataProfilesReply`` reply.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(GetMetadataProfilesCommand, chain)


@Message.define("command/metadata/profiles/reply")
class GetMetadataProfilesReply(CommandReply):
    """
    Reply to ``GetMetadataProfilesCommand``.

    Args:
        profiles: List of all global profiles.
    """

    profiles: MetadataProfileContainerList = dataclasses.field(default_factory=list)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: GetMetadataProfilesCommand,
        *,
        profiles: MetadataProfileContainerList,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            GetMetadataProfilesReply, cmd, success, message, profiles=profiles
        )

File: src/common/web/api/metadata/MetadataCommands.ts
import { Type } from "class-transformer";

import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";
import { MetadataProfileContainer, type MetadataProfileContainerList } from "../../data/entities/metadata/MetadataProfileContainer";

/**
 * Command to fetch all global metadata profiles. Requires a ``GetMetadataProfilesReply`` reply.
 */
@Message.define("command/metadata/profiles")
export class GetMetadataProfilesCommand extends Command {
    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<GetMetadataProfilesCommand> {
        return messageBuilder.buildCommand(GetMetadataProfilesCommand, {}, chain);
    }
}

/**
 * Reply to ``GetMetadataProfilesCommand``.
 *
 * @param profiles - List of all global profiles.
 */
@Message.define("command/metadata/profiles/reply")
export class GetMetadataProfilesReply extends CommandReply {
    // @ts-ignore
    @Type(() => MetadataProfileContainer)
    public readonly profiles: MetadataProfileContainerList = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: GetMetadataProfilesCommand,
        profiles: MetadataProfileContainerList,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<GetMetadataProfilesReply> {
        return messageBuilder.buildCommandReply(GetMetadataProfilesReply, cmd, success, message, { profiles: profiles });
    }
}
